File: MensLacrosseStats/View Controllers/Export/INSOEmailStatsViewController.m
//
//  INSOEmailStatsViewController.m
//  MensLacrosseStats
//
//  Created by James Dabrowski on 11/25/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

@import MessageUI;

#import "UIColor+INSOScorebookColor.h"

#import "INSOEmailStatsViewController.h"
#import "INSOEmailStatsFileGenerator.h"
#import "INSOProductManager.h"
#import "INSOMensLacrosseStatsConstants.h"

#import "Game.h"

@interface INSOEmailStatsViewController () <UITableViewDataSource, UITabBarDelegate, MFMailComposeViewControllerDelegate>
// IBOutlets
@property (weak, nonatomic) IBOutlet UISwitch *gameSummarySwitch;
@property (weak, nonatomic) IBOutlet UISwitch *playerStatsSwitch;
@property (weak, nonatomic) IBOutlet UISwitch *maxPrepsSwitch; 

@property (weak, nonatomic) IBOutlet UILabel *instructionLabel; 
@property (weak, nonatomic) IBOutlet UILabel *messageLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;

@property (weak, nonatomic) IBOutlet UIButton *exportStatsButton;

// IBActions
- (IBAction)done:(id)sender;
- (IBAction)toggledSwitch:(id)sender;
- (IBAction)exportStats:(id)sender;

@property (nonatomic, assign) BOOL isPreparingForBoys;

@property (nonatomic) NSData *maxPrepsAttachmentData;
@property (nonatomic) NSData *playerStatsAttachmentData;
@property (nonatomic) NSData *gameSummaryAttachmentData;

@end

@implementation INSOEmailStatsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.tableView.alwaysBounceVertical = NO;
    
// FIXME Which product are we exporting for?
    // Are we sending boys or girls stats?
//    if ([[[INSOProductManager sharedManager] appProductName] isEqualToString:INSOMensProductName]) {
        self.isPreparingForBoys = YES;
//    } else {
//        self.isPreparingForBoys = NO;
//    }
    
    [self.gameSummarySwitch setOn:[[NSUserDefaults standardUserDefaults] boolForKey:INSOExportGameSummaryDefaultKey]];
    [self.playerStatsSwitch setOn:[[NSUserDefaults standardUserDefaults] boolForKey:INSOExportPlayerStatsDefaultKey]];
    [self.maxPrepsSwitch setOn:[[NSUserDefaults standardUserDefaults] boolForKey:INSOExportMaxPrepsDefaultKey]];
    
    self.exportStatsButton.enabled = [self shouldEnableExportStatsButton];
    
    self.messageLabel.text = nil;

    if (![MFMailComposeViewController canSendMail]) {
        // Can't send email so disable UI and put up a message
        [self disableUI];
        [self.activityIndicator stopAnimating];
        self.messageLabel.text = @"Unable to export stats via email. Check your mail settings and try again.";
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - IBActions
- (void)done:(id)sender
{
    // Just dismiss
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)toggledSwitch:(id)sender
{
    self.exportStatsButton.enabled = [self shouldEnableExportStatsButton];
    
    [[NSUserDefaults standardUserDefaults] setBool:self.gameSummarySwitch.isOn forKey:INSOExportGameSummaryDefaultKey];
    [[NSUserDefaults standardUserDefaults] setBool:self.playerStatsSwitch.isOn forKey:INSOExportPlayerStatsDefaultKey];
    [[NSUserDefaults standardUserDefaults] setBool:self.maxPrepsSwitch.isOn forKey:INSOExportMaxPrepsDefaultKey];
}

- (void)exportStats:(id)sender
{
    // Freeze the UI
    [self disableUI];
    
    // Prepare the stats file(s)
    INSOEmailStatsFileGenerator* fileGenerator = [[INSOEmailStatsFileGenerator alloc] initWithGame:self.game];
    
    if (self.maxPrepsSwitch.isOn) {
        // Boys or girls?
        if (self.isPreparingForBoys) {
            [fileGenerator createMaxPrepsGameStatsData:^(NSData *maxPrepsData) {
                self.maxPrepsAttachmentData = maxPrepsData;
            }];
        } else {
            [fileGenerator createMaxPrepsGameStatsData:^(NSData *maxPrepsData) {
                self.maxPrepsAttachmentData = maxPrepsData;
            }];
        }
    }
    
    if (self.gameSummarySwitch.isOn) {
        [fileGenerator createGameSummaryData:^(NSData *gameSummaryData) {
            self.gameSummaryAttachmentData = gameSummaryData;
        }];
    }
    
    if (self.playerStatsSwitch.isOn) {
        [fileGenerator createPlayerStatsData:^(NSData *playerStatsData) {
            self.playerStatsAttachmentData = playerStatsData;
        }];
    }
    
    // Send the email
    [self actuallySendEmail];
}

#pragma mark - Private Methods
- (void)disableUI {
    self.gameSummarySwitch.enabled = NO;
    self.playerStatsSwitch.enabled = NO;
    self.maxPrepsSwitch.enabled = NO;
    self.exportStatsButton.enabled = NO;
    [self.activityIndicator startAnimating];
}

- (void)enableUI {
    self.gameSummarySwitch.enabled = YES;
    self.playerStatsSwitch.enabled = YES;
    self.maxPrepsSwitch.enabled = YES;
    self.exportStatsButton.enabled = YES;
    [self.activityIndicator stopAnimating];
}

- (BOOL)shouldEnableExportStatsButton
{
    return [MFMailComposeViewController canSendMail] && (self.gameSummarySwitch.isOn || self.playerStatsSwitch.isOn || self.maxPrepsSwitch.isOn);
}

- (void)actuallySendEmail
{
    if ([MFMailComposeViewController canSendMail]) {
        MFMailComposeViewController* mailViewcontroller = [[MFMailComposeViewController alloc] init];
        mailViewcontroller.mailComposeDelegate = self;
        
        // Get subject
        NSString* subject = [self mailMessageSubject];
        
        // Get body
        NSString* body = [self mailMessageBody];
        
         // Now set these things
        [mailViewcontroller setSubject:subject];
        [mailViewcontroller setMessageBody:body isHTML:NO];
        
        // Add attachment(s)
        if (self.maxPrepsAttachmentData) {
            NSString* fileName = NSLocalizedString(@"MaxPreps Export.txt", nil);
            [mailViewcontroller addAttachmentData:self.maxPrepsAttachmentData mimeType:@"text/txt" fileName:fileName];
        }
        
        if (self.gameSummaryAttachmentData) {
            NSString* fileName = NSLocalizedString(@"Game Summary.pdf", nil);
            [mailViewcontroller addAttachmentData:self.gameSummaryAttachmentData mimeType:@"application/pdf" fileName:fileName];
        }
        
        if (self.playerStatsAttachmentData) {
            NSString* fileName = NSLocalizedString(@"Player Stats.csv", nil);
            [mailViewcontroller addAttachmentData:self.playerStatsAttachmentData mimeType:@"text/csv" fileName:fileName];
        }

        if (self.isPreparingForBoys) {
            [mailViewcontroller.navigationBar setTintColor:[UIColor scorebookBlue]];
        } else {
            [mailViewcontroller.navigationBar setTintColor:[UIColor scorebookTeal]];
        }
        
        // Display the view to mail the message.
        [self presentViewController:mailViewcontroller animated:YES completion:^{
            // Re-enable the UI so that when the user dismisses the mail view,
            // the UI is ready to re-use.
            [self enableUI];
        }];
        
    } else {
        // unable to send mail. Hmmm;
        NSLog(@"Error - Unable to send email. Should never have gotten here.");
    }
}

- (NSString*)mailMessageSubject
{
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterShortStyle];
    
    NSString* localizedSubjectString = NSLocalizedString(@"Lacrosse Stats export for %@ vs. %@ on %@", nil);
    NSString* subject = [NSString stringWithFormat:localizedSubjectString, self.game.visitingTeam, self.game.homeTeam, [dateFormatter stringFromDate:self.game.gameDateTime]];
    
    return subject;
}

- (NSString*)mailMessageBody
{
    NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterShortStyle];
    
    NSString* localizedMessageString = NSLocalizedString(@"Stats files for %@ vs. %@ on %@ at %@.\n\n", nil);
    NSString* messageBody = [NSString stringWithFormat:localizedMessageString, self.game.visitingTeam, self.game.homeTeam, [dateFormatter stringFromDate:self.game.gameDateTime], self.game.location];
    NSMutableString *mutableMessageBody = [[NSMutableString alloc] initWithString:messageBody];
    
    int fileCount = self.gameSummarySwitch.isOn;
    fileCount += self.playerStatsSwitch.isOn;
    fileCount += self.maxPrepsSwitch.isOn;
    
    if (fileCount == 1) {
        [mutableMessageBody appendString:NSLocalizedString(@"The following file is attached:\n", nil)];
    } else {
        [mutableMessageBody appendString:NSLocalizedString(@"The following files are attached:\n", nil)];
    }
    
    if (self.gameSummarySwitch.isOn) {
        [mutableMessageBody appendString:NSLocalizedString(@"Game Summary", nil)];
        [mutableMessageBody appendString:@"\n"];
    }
    
    if (self.playerStatsSwitch.isOn) {
        [mutableMessageBody appendString:NSLocalizedString(@"Individual player stats", nil)];
        [mutableMessageBody appendString:@"\n"];
    }
    
    if (self.maxPrepsSwitch.isOn) {
        [mutableMessageBody appendString:NSLocalizedString(@"MaxPreps file", nil)];
        [mutableMessageBody appendString:@"\n"];
    }
    
    return mutableMessageBody;
}

#pragma mark - MFMailComposeControllerDelegate
- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
    switch (result) {
        case MFMailComposeResultSent:
            // It would be nice if I could play a sound here.
            self.messageLabel.text = NSLocalizedString(@"Email sent.", nil);
            break;
        case MFMailComposeResultSaved:
            self.messageLabel.text = NSLocalizedString(@"Draft email saved.", nil);
            break;
        case MFMailComposeResultCancelled:
            self.messageLabel.text = NSLocalizedString(@"Email cancelled.", nil);
            break;
        case MFMailComposeResultFailed:
            self.messageLabel.text = NSLocalizedString(@"An error occurred when trying to compose this email.", nil);
            break;
        default:
            self.messageLabel.text = nil;
            break;
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Common/ViewControllers/Export/INSOEmailStatsViewController.swift
//  Converted to Swift 5.5 by Swiftify v5.5.27463 - https://swiftify.com/
//
//  INSOEmailStatsViewController.swift
//  MensLacrosseStats
//
//  Created by James Dabrowski on 11/25/15.
//  Copyright © 2015 Intangible Software. All rights reserved.
//

import MessageUI
import UIKit

class INSOEmailStatsViewController: UITableViewController, UITabBarDelegate, MFMailComposeViewControllerDelegate {
    var game: Game?
    
    private var maxPrepsAttachmentData: Data?
    private var playerStatsAttachmentData: Data?
    private var gameSummaryAttachmentData: Data?

    // MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.alwaysBounceVertical = false

        gameSummarySwitch.isOn = UserDefaults.standard.bool(forKey: INSOExportGameSummaryDefaultKey)
        playerStatsSwitch.isOn = UserDefaults.standard.bool(forKey: INSOExportPlayerStatsDefaultKey)
        maxPrepsSwitch.isOn = UserDefaults.standard.bool(forKey: INSOExportMaxPrepsDefaultKey)

        exportStatsButton.isEnabled = shouldEnableExportStatsButton()

        messageLabel.text = nil

        if !MFMailComposeViewController.canSendMail() {
            // Can't send email so disable UI and put up a message
            disableUI()
            activityIndicator.stopAnimating()
            messageLabel.text = "Unable to export stats via email. Check your mail settings and try again."
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - IBOutlets
    @IBOutlet private weak var gameSummarySwitch: UISwitch!
    @IBOutlet private weak var playerStatsSwitch: UISwitch!
    @IBOutlet private weak var maxPrepsSwitch: UISwitch!
    @IBOutlet private weak var instructionLabel: UILabel!
    @IBOutlet private weak var messageLabel: UILabel!
    @IBOutlet private weak var activityIndicator: UIActivityIndicatorView!
    @IBOutlet private weak var exportStatsButton: UIButton!
    
    // MARK: - IBActions

    @IBAction func done(_ sender: Any?) {
        // Just dismiss
        presentingViewController?.dismiss(animated: true)
    }

    @IBAction func toggledSwitch(_ sender: Any?) {
        exportStatsButton.isEnabled = shouldEnableExportStatsButton()

        UserDefaults.standard.set(gameSummarySwitch.isOn, forKey: INSOExportGameSummaryDefaultKey)
        UserDefaults.standard.set(playerStatsSwitch.isOn, forKey: INSOExportPlayerStatsDefaultKey)
        UserDefaults.standard.set(maxPrepsSwitch.isOn, forKey: INSOExportMaxPrepsDefaultKey)
    }

    @IBAction func exportStats(_ sender: Any?) {
        guard let game = game else {
            // Disable the UI and be done if we don't have a game object for some reason.
            disableUI()
            return
        }
        
        guard let fileGenerator = INSOEmailStatsFileGenerator(game: game) else {
            disableUI()
            return
        }
        
        // Freeze the UI
        disableUI()
        
        // Clear any old data files hanging around
        maxPrepsAttachmentData = nil
        gameSummaryAttachmentData = nil
        playerStatsAttachmentData = nil

        // Now generate the attachments. 
        if maxPrepsSwitch.isOn {
            fileGenerator.createMaxPrepsGameStatsData { self.maxPrepsAttachmentData = $0 }
        }

        if gameSummarySwitch.isOn {
            fileGenerator.createGameSummaryData { self.gameSummaryAttachmentData = $0 }
        }

        if playerStatsSwitch.isOn {
            fileGenerator.createPlayerStatsData { self.playerStatsAttachmentData = $0 }
        }

        // Send the email
        actuallySendEmail()
    }

    // MARK: - Private Methods
    func disableUI() {
        gameSummarySwitch.isEnabled = false
        playerStatsSwitch.isEnabled = false
        maxPrepsSwitch.isEnabled = false
        exportStatsButton.isEnabled = false
        activityIndicator.startAnimating()
    }

    func enableUI() {
        gameSummarySwitch.isEnabled = true
        playerStatsSwitch.isEnabled = true
        maxPrepsSwitch.isEnabled = true
        exportStatsButton.isEnabled = true
        activityIndicator.stopAnimating()
    }

    func shouldEnableExportStatsButton() -> Bool {
        return MFMailComposeViewController.canSendMail() && (gameSummarySwitch.isOn || playerStatsSwitch.isOn || maxPrepsSwitch.isOn)
    }

    func actuallySendEmail() {
        if MFMailComposeViewController.canSendMail() {
            let mailViewcontroller = MFMailComposeViewController()
            mailViewcontroller.mailComposeDelegate = self

            // Set subject and body
            mailViewcontroller.setSubject(mailMessageSubject())
            mailViewcontroller.setMessageBody(mailMessageBody(), isHTML: false)

            // Add attachment(s)
            if let maxPrepsAttachmentData = maxPrepsAttachmentData {
                let fileName = NSLocalizedString("MaxPreps Export.txt", comment: "")
                mailViewcontroller.addAttachmentData(maxPrepsAttachmentData, mimeType: "text/txt", fileName: fileName)
            }

            if let gameSummaryAttachmentData = gameSummaryAttachmentData {
                let fileName = NSLocalizedString("Game Summary.pdf", comment: "")
                mailViewcontroller.addAttachmentData(gameSummaryAttachmentData, mimeType: "application/pdf", fileName: fileName)
            }

            if let playerStatsAttachmentData = playerStatsAttachmentData {
                let fileName = NSLocalizedString("Player Stats.csv", comment: "")
                mailViewcontroller.addAttachmentData(playerStatsAttachmentData, mimeType: "text/csv", fileName: fileName)
            }

            if Target.isMens {
                mailViewcontroller.navigationBar.tintColor = UIColor.scorebookBlue()
            } else if Target.isWomens {
                mailViewcontroller.navigationBar.tintColor = UIColor.scorebookTeal()
            }

            // Display the view to mail the message.
            present(mailViewcontroller, animated: true) { [self] in
                // Re-enable the UI so that when the user dismisses the mail view,
                // the UI is ready to re-use.
                enableUI()
            }
        } else {
            // unable to send mail. Hmmm;
            print("Error - Unable to send email. Should never have gotten here.")
        }
    }

    func mailMessageSubject() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .short
        
        var subject = ""
        guard let game = game else {
            return subject
        }

        subject = NSLocalizedString("Lacrosse Stats export for %@ vs. %@ on %@", comment: "")
        if let gameDateTime = game.gameDateTime, let visitingTeam = game.visitingTeam, let homeTeam = game.homeTeam {
            subject = String(format: subject, visitingTeam, homeTeam, dateFormatter.string(from: gameDateTime))
        }

        return subject
    }

    func mailMessageBody() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .short

        var messageBody: String = ""
        guard let game = game else {
            return messageBody
        }
        messageBody = NSLocalizedString("Stats files for %@ vs. %@ on %@ at %@.\n\n", comment: "")
        
        if let gameDateTime = game.gameDateTime, let visitingTeam = game.visitingTeam, let homeTeam = game.homeTeam, let location = game.location {
            messageBody = String(format: messageBody, visitingTeam, homeTeam, dateFormatter.string(from: gameDateTime), location)
        }

        var fileCount = 0
        if gameSummarySwitch.isOn { fileCount += 1 }
        if playerStatsSwitch.isOn { fileCount += 1 }
        if maxPrepsSwitch.isOn { fileCount += 1 }

        if fileCount == 1 {
            messageBody += NSLocalizedString("The following file is attached:\n", comment: "")
        } else {
            messageBody += NSLocalizedString("The following files are attached:\n", comment: "")
        }

        if gameSummarySwitch.isOn {
            messageBody += NSLocalizedString("Game Summary", comment: "")
            messageBody += "\n"
        }

        if playerStatsSwitch.isOn {
            messageBody += NSLocalizedString("Individual player stats", comment: "")
            messageBody += "\n"
        }

        if maxPrepsSwitch.isOn {
            messageBody += NSLocalizedString("MaxPreps file", comment: "")
            messageBody += "\n"
        }

        return messageBody
    }

    // MARK: - MFMailComposeControllerDelegate

    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        switch result {
        case .sent:
            // It would be nice if I could play a sound here.
            messageLabel.text = NSLocalizedString("Email sent.", comment: "")
        case .saved:
            messageLabel.text = NSLocalizedString("Draft email saved.", comment: "")
        case .cancelled:
            messageLabel.text = NSLocalizedString("Email cancelled.", comment: "")
        case .failed:
            messageLabel.text = NSLocalizedString("An error occurred when trying to compose this email.", comment: "")
        default:
            messageLabel.text = nil
        }

        dismiss(animated: true)
    }
}
